google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerCloudMonitoringExporter.java

```diff
@@ -0,0 +1,246 @@
+      return CompletableResultCode.ofFailure();
+    }
+
+    List<TimeSeries> spannerTimeSeries;
+    try {
+      spannerTimeSeries =
+          SpannerCloudMonitoringExporterUtils.convertToSpannerTimeSeries(spannerMetricData);
+    } catch (Throwable e) {
+      logger.log(
+          Level.WARNING,
+          "Failed to convert spanner metric data to cloud monitoring timeseries.",
+          e);
+      return CompletableResultCode.ofFailure();
+    }
+
+    ProjectName projectName = ProjectName.of(spannerProjectId);
+
+    ApiFuture<List<Empty>> futureList = exportTimeSeriesInBatch(projectName, spannerTimeSeries);
+
+    CompletableResultCode spannerExportCode = new CompletableResultCode();
+    ApiFutures.addCallback(
+        futureList,
+        new ApiFutureCallback<List<Empty>>() {
+          @Override
+          public void onFailure(Throwable throwable) {
+            if (spannerExportFailureLogged.compareAndSet(false, true)) {
+              String msg = "createServiceTimeSeries request failed for spanner metrics.";
+              if (throwable instanceof PermissionDeniedException) {
+                // TODO: Add the link of public documentation when available in the log message.
+                msg +=
+                    String.format(
+                        " Need monitoring metric writer permission on project=%s.",
+                        projectName.getProject());
+              }
+              logger.log(Level.WARNING, msg, throwable);
+            }
+            spannerExportCode.fail();
+          }
+
+          @Override
+          public void onSuccess(List<Empty> empty) {
+            // When an export succeeded reset the export failure flag to false so if there's a
+            // transient failure it'll be logged.
+            spannerExportFailureLogged.set(false);
+            spannerExportCode.succeed();
+          }
+        },
+        MoreExecutors.directExecutor());
+
+    return spannerExportCode;
+  }
+
+  private ApiFuture<List<Empty>> exportTimeSeriesInBatch(
+      ProjectName projectName, List<TimeSeries> timeSeries) {
+    List<ApiFuture<Empty>> batchResults = new ArrayList<>();
+
+    for (List<TimeSeries> batch : Iterables.partition(timeSeries, EXPORT_BATCH_SIZE_LIMIT)) {
+      CreateTimeSeriesRequest req =
+          CreateTimeSeriesRequest.newBuilder()
+              .setName(projectName.toString())
+              .addAllTimeSeries(batch)
+              .build();
+      ApiFuture<Empty> f = this.client.createServiceTimeSeriesCallable().futureCall(req);
+      batchResults.add(f);
+    }
+
+    return ApiFutures.allAsList(batchResults);
+  }
+
+  @Override
+  public CompletableResultCode flush() {
+    if (lastExportCode != null) {
+      return lastExportCode;
+    }
+    return CompletableResultCode.ofSuccess();
+  }
+
+  @Override
+  public CompletableResultCode shutdown() {
+    if (!isShutdown.compareAndSet(false, true)) {
```

Comment:
@psx95 do we do this in our user facing exporter?

Usually the user would not call this directly but calling `shutdown()` on `OpenTelemetrySdk` or `MeterProvider` would run this for you, handling the duplicate calls.

Reply:
No, we do not have this check. 

This is because we rely on the `shutdown()` method of the [MetricServiceClient](https://github.com/googleapis/google-cloud-java/blob/5a2e219452e1f57e2c26c8bd4edd2d8311d0a849/java-monitoring/google-cloud-monitoring/src/main/java/com/google/cloud/monitoring/v3/MetricServiceClient.java#L2097). This class implements [`BackgroundResource`](https://github.com/googleapis/sdk-platform-java/blob/502f80101dec191befb660a1aba6d0c354758c18/gax-java/gax/src/main/java/com/google/api/gax/core/BackgroundResource.java#L38) and the documentation suggests invocation has no effect if already shutdown, so we don't care if its called multiple times.

IIUC, a similar thing can be done in this exporter.